src/EventSubscriber/SmsMessageProcessor.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+namespace Drupal\sms\EventSubscriber;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Url;
+use Drupal\sms\Entity\SmsGatewayInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\sms\Event\RecipientGatewayEvent;
+use Drupal\sms\Event\SmsMessageEvent;
+use Drupal\sms\Exception\RecipientRouteException;
+use Drupal\sms\Entity\SmsGateway;
+
+/**
+ * Handles messages before they are processed by queue(), send(), or incoming().
+ *
+ * Messages queued via queue() are destined for send() or incoming(), they will
+ * not be double processed.
+ */
+class SmsMessageProcessor implements EventSubscriberInterface {
+
+  /**
+   * The event dispatcher.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $eventDispatcher;
+
+  /**
+   * The configuration factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * Creates a new SmsMessageProcessor controller.
+   *
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $event_dispatcher
+   *   The event dispatcher.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The configuration factory.
+   */
+  public function __construct(EventDispatcherInterface $event_dispatcher, ConfigFactoryInterface $config_factory) {
+    $this->eventDispatcher = $event_dispatcher;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * Ensure all recipients are routed to a gateway.
+   *
+   * Messages will be split into multiple if recipients need to be routed to
+   * different gateways.
+   *
+   * @param \Drupal\sms\Event\SmsMessageEvent $event
+   *   The SMS message preprocess event.
+   *
+   * @throws \Drupal\sms\Exception\RecipientRouteException
+   *   Guarantees a gateway is set on the message, otherwise this exception is
+   *   thrown.
+   */
+  public function ensureGateways(SmsMessageEvent $event) {
+    $sms_messages = $event->getMessages();
+    $result = [];
+
+    // Ignore messages if they already have a gateway.
+    foreach ($sms_messages as $k => $sms_message) {
+      if ($sms_message->getGateway() instanceof SmsGatewayInterface) {
+        unset($sms_messages[$k]);
+        $result[] = $sms_message;
+      }
+    }
+
+    // Ensure all recipients in this message can be routed to a gateway.
+    foreach ($sms_messages as $sms_message) {
+      $gateways = [];
+
+      $recipients_all = $sms_message->getRecipients();
+      foreach ($recipients_all as $recipient) {
+        $gateway = $this->getGatewayForPhoneNumber($recipient);
```

Comment:
To clarify, the slowdown will be with the amount of computation that an event subscriber will need to determine gateway for one recipient vs. all recipients. It's much less than O(n) so it'll be faster if all are calculated at once.